util/src/main/java/io/camunda/zeebe/util/startup/StartupProcess.java

```diff
@@ -0,0 +1,292 @@
+
+  private boolean startupCalled = false;
+  private ActorFuture<CONTEXT> shutdownFuture;
+  private ActorFuture<CONTEXT> startupFuture;
+
+  /**
+   * Constructs the startup process
+   *
+   * @param steps the steps to execute; must not be {@code null}
+   */
+  public StartupProcess(final List<StartupStep<CONTEXT>> steps) {
+    this(LoggerFactory.getLogger("io.camunda.zeebe.util.startup"), steps);
+  }
+
+  /**
+   * Constructs the startup process
+   *
+   * @param logger the logger to use for messages related to the startup process; must not be {@code
+   *     null}
+   * @param steps the steps to execute; must not be {@code null}
+   */
+  public StartupProcess(final Logger logger, final List<StartupStep<CONTEXT>> steps) {
+    this.steps = new ArrayDeque<>(Objects.requireNonNull(steps));
+    this.logger = Objects.requireNonNull(logger);
+  }
+
+  /**
+   * Executes the startup logic
+   *
+   * @param concurrencyControl the startup context at the start of this step
+   * @param context the startup context at the start of this step
+   * @return future with startup context at the end of this step
+   */
+  public ActorFuture<CONTEXT> startup(
+      final ConcurrencyControl concurrencyControl, final CONTEXT context) {
+
+    final var result = concurrencyControl.<CONTEXT>createFuture();
+    concurrencyControl.submit(() -> startupSynchronized(concurrencyControl, context, result));
+
+    return result;
+  }
+
+  /**
+   * Executes the shutdown logic
+   *
+   * @param context the shutdown context at the start of this step
+   * @return future with the shutdown context at the end of this step.
+   */
+  public ActorFuture<CONTEXT> shutdown(
+      final ConcurrencyControl concurrencyControl, final CONTEXT context) {
+    final var result = concurrencyControl.<CONTEXT>createFuture();
+    concurrencyControl.submit(() -> shutdownSynchronized(concurrencyControl, context, result));
+
+    return result;
+  }
+
+  private void startupSynchronized(
+      final ConcurrencyControl concurrencyControl,
+      final CONTEXT context,
+      final ActorFuture<CONTEXT> startupFuture) {
+    logger.debug("Startup was called with context: " + context);
+    if (startupCalled) {
+      throw new IllegalStateException("startup(...) must only be called once");
+    }
+    startupCalled = true;
+    this.startupFuture = startupFuture;
+
+    // reset future when we are done
+    concurrencyControl.runOnCompletion(startupFuture, (result, error) -> this.startupFuture = null);
+
+    final var stepsToStart = new ArrayDeque<>(steps);
+
+    logger.info("Starting startup process");
+    proceedWithStartupSynchronized(concurrencyControl, stepsToStart, context, startupFuture);
+  }
+
+  private void proceedWithStartupSynchronized(
+      final ConcurrencyControl concurrencyControl,
+      final Queue<StartupStep<CONTEXT>> stepsToStart,
+      final CONTEXT context,
+      final ActorFuture<CONTEXT> startupFuture) {
+    if (stepsToStart.isEmpty()) {
```

Comment:
:question: :thought_balloon: I typically model these as guard clauses, i.e. with a return statement. So:

```java
if (someGuardCondition) {
  // do something
  return;
}

if (somethingElse) {
  // do more
  return;
}

// common case/more complex business logic
```

I find you then have a clear separation between the common/complex case and the others, which I find easier to read. I'm more curious here to hear your thoughts here (or also fine if you hadn't really thought about it :smile:)

Reply:
Have not given it too much thought, to be honest. I use guard clauses more for errors or exceptions, not for ordinary states.

I also tend to use guard clauses more when the list of guards gets longer. 

Since this is an uncommon way to implement recursion, I like to see the end condition for the recursion in the beginning. And I wouldn't use a guard clause for this.

Anyway, I am happy to change it if you think the other way is better.

Reply:
I don't have concrete proof that it's any better than a personal preference - I find it easier to read because I can quickly read and then dismiss the non-common case - but is that true for everyone? Does it actually make it more readable generally? The other way is also readable, so the benefits would probably just be marginal. As much as I like bike shedding (clearly :smile:), both approaches are readable imo, so :man_shrugging: